### ISSUE TITLE
Logger integration

### ISSUE THREAD

**Issue by jimrobinson** [confluentinc/confluent-kafka-go#54]
I ran a test to see what happened when my application encountered a broker that was shut down.  Instead of getting back an errors from kafka.NewProducer or from setting a (*kafka.Message).TopicPartition.Error on message submission, what I was seeing was that the library appears to be writing directly to stderr.  An example:

```
%3|1490436188.086|ERROR|kpush#producer-1| [thrd:localhost:9092/bootstrap]: localhost:9092/bootstrap: Connect to ipv6#[::1]:9092 failed: Connection refused
%3|1490436188.086|ERROR|kpush#producer-1| [thrd:localhost:9092/bootstrap]: 1/1 brokers are down
```

Is there some way to control this behavior?  I didn't see any options for controlling the logging output in

https://github.com/confluentinc/confluent-kafka-go
http://docs.confluent.io/3.2.0/clients/confluent-kafka-go/index.html

**Comment by edenhill** [comment]
Broker connection errors are considered temporary and the client will try to recover automatically, that's why these errors are not directly propagated as per-message errors but rather global errors (or logs).
The not-yet-produced messages will remain in internal queues until the relevant broker(s) comes up again or the messages time out (`message.timeout.ms`, def 5 minutes).


Regarding the logging output: it is not currently possible but it is on the roadmap.
Do you have a preference for how you would like to receive log messages? A Log event on the .Events channel? A separate channel? something else?

**Comment by jimrobinson** [comment]
Hi,

I think a separate "notification" log channel would be great.  Otherwise possibly allow someone the option of the librdkafka library just calling the standard logger in https://golang.org/pkg/log/, thereby letting them handle sending the log to a file or whatever they want via that package's options.

**Comment by tchap** [comment]
Hi, I vote for a separate channel to receive a structured as well-defined errors, no strings please.

**Comment by eliaslevy** [comment]
May also want to give the option to just configure the log callback, which can be done in `librdkafka` via `rd_kafka_conf_set_log_cb`.  The current lack of control over log output is unworkable in a number of environments.

**Comment by edenhill** [comment]
That is somewhat problematic since the librdkafka log_cb may be called from internal librdkafka threads, it might be safer to emit log messages as Events and let the application poll/consume them.
This is what we did in the Python client.

**Comment by eliaslevy** [comment]
I wonder about possible issues with that approach, as it would delay the logging of messages, which may be lot in the case of some fatal failure in the meantime.  But it would certainly be an improvement on the current state.

**Comment by tchap** [comment]
A separate channel perhaps? It would have to enabled (by default disabled) using a special config key I guess so that the producer/consumer does not deadlock when the channel is not being received upon.

**Comment by eliaslevy** [comment]
That could work.  A separate go routine can consume the log messages.

**Comment by scraly** [comment]
Is it some news about this point?
In a golang app using confluent-kafka-go, how to handle with connection to broker & librdkafka err logs?
Thx

**Comment by edenhill** [comment]
This is on our roadmap, we're aiming to add proper logging support later this year.
PRs are of course welcome before that :)
Corresponding Python client's logger, start here:
https://github.com/confluentinc/confluent-kafka-python/blob/master/confluent_kafka/src/confluent_kafka.c#L1669

**Comment by cshunger** [comment]
@onorua I upgraded my kafka client to v0.11.6 and this was not the case for the consumer. 

For the consumer when:
```
ev := k.Consumer.Poll(100)
if ev == nil {
    continue
}
```
is invoked while the brokers are down, the logs
```
%3|1544122290.050|FAIL|metrics-publisher-service-local#consumer-1| [thrd:localkafka:9092/bootstrap]: localkafka:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused
%3|1544122290.050|ERROR|metrics-publisher-service-local#consumer-1| [thrd:localkafka:9092/bootstrap]: localkafka:9092/bootstrap: Connect to ipv4#127.0.0.1:9092 failed: Connection refused
%3|1544122290.050|ERROR|metrics-publisher-service-local#consumer-1| [thrd:localkafka:9092/bootstrap]: 1/1 brokers are down
```
 are printed and `ev` is `nil`. Should a new issue be opened for the consumer?

**Comment by edenhill** [comment]
v0.11.6 added proper Error reporting, but there is still no support for specifying an alternate logger. All logs currently go to stderr.

We're scheduling this for early 2019, unless someone from the community wants to help out.

This boils down to enabling librdkafka log queue and modifying the event loop in event.go to translate librdkafka log events to the Go log framework.

**Comment by PChou** [comment]
I implement this recently, and use  both separate goroutine and channel. https://github.com/confluentinc/confluent-kafka-go/pull/295

**Comment by rigelbm** [comment]
Merged in #421.
